vmm/src/sigsys_handler.rs

```diff
@@ -35,17 +41,66 @@ extern "C" fn sigsys_handler(
     if num != libc::SIGSYS {
         return;
     }
-    let syscall = unsafe { *(info as *const i32).offset(SI_OFF_SYSCALL) as usize };
-    METRICS.seccomp.num_faults.inc();
-    METRICS.seccomp.bad_syscalls[syscall].inc();
+
+    // We're also trying to log something here, which might not be re-entrant (now, or at some
+    // point in the future), so we make sure this logic only runs once. If another signal causes
+    // the handler to run again, we simply exit the process with the appropriate error code. This
+    // should be extremely unlikely to happen.
+    if HANDLING_BAD_SYSCALL.compare_and_swap(false, true, Ordering::SeqCst) == false {
+        let syscall = unsafe { *(info as *const i32).offset(SI_OFF_SYSCALL) as usize };
+        METRICS.seccomp.num_faults.inc();
+        METRICS.seccomp.bad_syscalls[syscall].inc();
+        error!("Shutting down VM after intercepting a bad syscall.");
+        // Log the metrics before exiting.
+        if let Err(e) = LOGGER.log_metrics() {
+            error!("Failed to log metrics while stopping: {}", e);
+        }
+    }
+
+    // Safe because we're terminating the process anyway. We don't actually do anything when
+    // running unit tests.
+    #[cfg(not(test))]
+    unsafe {
+        libc::_exit(super::FC_EXIT_CODE_BAD_SYSCALL as i32)
+    };
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
-    use seccomp::{setup_seccomp, SeccompLevel};
+
     use std::process;
 
+    use seccomp::{setup_seccomp, SeccompLevel};
+
+    // This function is used when running unit tests, so all the unsafes are safe.
+    fn cpu_count() -> usize {
+        use std::mem;
```

Comment:
nit: group the `use`s in the module root instead of splitting them per function (ex. `std::process` isn't used inside `cpu_count`, but it's not imported in-place just in the function that needs it)